framework/include/userobject/ElementLoopUserObject.h

```diff
@@ -0,0 +1,89 @@
+#ifndef ELEMENTLOOPUSEROBJECT_H
+#define ELEMENTLOOPUSEROBJECT_H
+
+#include "GeneralUserObject.h"
+#include "ThreadedElementLoop.h"
+#include "Coupleable.h"
+#include "MooseVariableInterface.h"
+#include "ScalarCoupleable.h"
+#include "MooseVariableDependencyInterface.h"
+#include "ZeroInterface.h"
+#include "MooseVariable.h"
+#include "SubProblem.h"
+#include "NonlinearSystem.h"
+#include "Assembly.h"
+#include "FEProblem.h"
+#include "MooseMesh.h"
+
+#include "libmesh/elem.h"
+#include "libmesh/parallel_algebra.h"
+
+// Forward Declarations
+class ElementLoopUserObject;
+
+namespace libMesh
+{
+class Elem;
+class QBase;
+}
+
+template<>
+InputParameters validParams<ElementLoopUserObject>();
+
+/**
+ * This will loop over elements and do things
+ *
+ * !!! EXPERIMENTAL !!! Just to see what it takes to do such a thing. Not meant for real use ;-)
+ */
+class ElementLoopUserObject :
+    public GeneralUserObject,
+    public ThreadedElementLoop<ConstElemRange>,
```

Comment:
This is a crucial part of the design. What is needed for rDG is several sweeps through mesh. One sweep for gradient reconstruction and one for slope limiting. We cannot do that in `ElementUserObject`, because all those are executed at once. Since `GeneralUserObjects` are execute "linearly", we used that feature to chain the sweeps. I did not want another system in MOOSE for that 😃